### PR TITLE
[FLINK-28610][runtime] Enable speculative execution for sources

### PR DESCRIPTION
**Pull request by zhuzhurk**
## What is the purpose of the change

This PR enables speculative execution for sources and added IT cases for source speculation.

## Verifying this change

  - *Added integration tests in SpeculativeSchedulerITCase*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java [hunk 1]
```diff
@@ -109,14 +132,35 @@ void testSpeculativeExecutionWithAdaptiveParallelism() throws Exception {
...
+        JobClient client = executeJobAsync(configuration, this::setupJobWithSlowMap);
 
         assertThatThrownBy(
                         () -> client.getJobExecutionResult().get(30, TimeUnit.SECONDS),
                         "The local node is expected to be blocked but it is not.")
                 .isInstanceOf(TimeoutException.class);
     }
 
+    @Test
+    public void testSpeculativeExecutionOfSourceFunction() throws Exception {
```
Comment (wanglijie95) [inline flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java:144]: public is not needed

File: flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java [hunk 2]
```diff
@@ -109,14 +132,35 @@ void testSpeculativeExecutionWithAdaptiveParallelism() throws Exception {
...
+        executeJob(this::setupJobWithSlowSourceFunction);
+        waitUntilJobArchived();
+        checkResults();
+    }
+
+    @Test
+    public void testSpeculativeExecutionOfInputFormatSource() throws Exception {
```
Comment (wanglijie95) [inline flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java:151]: public is not needed

File: flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java [hunk 3]
```diff
@@ -109,14 +132,35 @@ void testSpeculativeExecutionWithAdaptiveParallelism() throws Exception {
...
+        executeJob(this::setupJobWithSlowInputFormatSource);
+        waitUntilJobArchived();
+        checkResults();
+    }
+
+    @Test
+    public void testSpeculativeExecutionOfNewSource() throws Exception {
```
Comment (wanglijie95) [inline flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java:158]: public is not needed

**Review by wanglijie95 — APPROVED** [review]
Thanks for creating this PR @zhuzhurk. LGTM. I left 3 minor comments.

**Comment by zhuzhurk** [comment]
@flinkbot run azure

**Comment by zhuzhurk** [comment]
@flinkbot run azure

**Comment by zhuzhurk** [comment]
@flinkbot run azure

**Pull request closed without merging**